presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -963,7 +963,9 @@ else if (criteria instanceof JoinOn) {
                 // we do it further down when after we determine which subexpressions apply to left vs right tuple)
                 ExpressionAnalyzer analyzer = ExpressionAnalyzer.create(analysis, session, metadata, sqlParser, accessControl);
 
-                Type clauseType = analyzer.analyze(expression, output);
+                // need to register coercions in case when join criteria requires coercion (e.g. join on char(1) = char(2))
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, output);
```

Comment:
Fair.
I've removed `canonicalizeExpression` from `visitJoin`, squashed all fixups and it's ready for you now. :)

I checked plans for some queries with canonicalized expressions:

Before patch:
```sql
presto:tiny> explain select count(*) > 0 from nation join region ON date('2014-12-12') = current_timestamp;
                       Query Plan
---------------------------------------------------------
 - Output[_col0] => [expr:boolean]
         _col0 := expr
     - Project[] => [expr:boolean]
             expr := ("count" > BIGINT '0')
         - LocalExchange[ROUND_ROBIN] () => count:bigint
             - Aggregate => [count:bigint]
                     count := "count"(*)
                 - Values => []

(1 row)
```

After patch:
```sql
presto:tiny> explain select count(*) > 0 from nation join region ON date('2014-12-12') = current_timestamp;
                       Query Plan
---------------------------------------------------------
 - Output[_col0] => [expr:boolean]
         _col0 := expr
     - Project[] => [expr:boolean]
             expr := ("count" > BIGINT '0')
         - LocalExchange[ROUND_ROBIN] () => count:bigint
             - Aggregate => [count:bigint]
                     count := "count"(*)
                 - Values => []

(1 row)
```
Both identical.

---

Before patch:
```sql
presto:tiny> explain select count(*) > 0 from nation join region ON true;
                                                  Query Plan
---------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [expr:boolean]
         _col0 := expr
     - Project[] => [expr:boolean]
             expr := ("count" > BIGINT '0')
         - LocalExchange[ROUND_ROBIN] () => count:bigint
             - Aggregate(FINAL) => [count:bigint]
                     count := "count"("count_7")
                 - LocalExchange[SINGLE] () => count_7:bigint
                     - RemoteExchange[GATHER] => count_7:bigint
                         - Aggregate(PARTIAL) => [count_7:bigint]
                                 count_7 := "count"(*)
                             - CrossJoin => []
                                 - TableScan[tpch:tpch:nation:sf0.01, originalConstraint = true] => []
                                 - LocalExchange[SINGLE] () =>
                                     - RemoteExchange[REPLICATE] =>
                                         - TableScan[tpch:tpch:region:sf0.01, originalConstraint = true] => []

(1 row)
```

After patch:
```sql
presto:tiny> explain select count(*) > 0 from nation join region ON CAST((CASE WHEN (TRUE IS NOT NULL) THEN 1.2 ELSE 1.2 END) as real) = CAST(1.2 as decimal(2,1));
                                                  Query Plan
---------------------------------------------------------------------------------------------------------------
 - Output[_col0] => [expr:boolean]
         _col0 := expr
     - Project[] => [expr:boolean]
             expr := ("count" > BIGINT '0')
         - LocalExchange[ROUND_ROBIN] () => count:bigint
             - Aggregate(FINAL) => [count:bigint]
                     count := "count"("count_7")
                 - LocalExchange[SINGLE] () => count_7:bigint
                     - RemoteExchange[GATHER] => count_7:bigint
                         - Aggregate(PARTIAL) => [count_7:bigint]
                                 count_7 := "count"(*)
                             - CrossJoin => []
                                 - TableScan[tpch:tpch:nation:sf0.01, originalConstraint = true] => []
                                 - LocalExchange[SINGLE] () =>
                                     - RemoteExchange[REPLICATE] =>
                                         - TableScan[tpch:tpch:region:sf0.01, originalConstraint = true] => []

(1 row)
```
Both identical.

---

Before patch:
```sql
presto:tiny> explain select count(*) > 0 from nation join region ON false;
                       Query Plan
---------------------------------------------------------
 - Output[_col0] => [expr:boolean]
         _col0 := expr
     - Project[] => [expr:boolean]
             expr := ("count" > BIGINT '0')
         - LocalExchange[ROUND_ROBIN] () => count:bigint
             - Aggregate => [count:bigint]
                     count := "count"(*)
                 - Values => []

(1 row)
```

After patch:
```sql
presto:tiny> explain select count(*) > 0 from nation join region ON CAST((CASE WHEN (TRUE IS NOT NULL) THEN 'a' ELSE 'a' END) as char(1)) = CAST('a' as char(2));
                       Query Plan
---------------------------------------------------------
 - Output[_col0] => [expr:boolean]
         _col0 := expr
     - Project[] => [expr:boolean]
             expr := ("count" > BIGINT '0')
         - LocalExchange[ROUND_ROBIN] () => count:bigint
             - Aggregate => [count:bigint]
                     count := "count"(*)
                 - Values => []

(1 row)
```

Both identical.

---

Of course if one used exactly the same query (instead of `on false`), then `explain select count(*) > 0 from nation join region ON CAST((CASE WHEN (TRUE IS NOT NULL) THEN 'a' ELSE 'a' END) as char(1)) = CAST('a' as char(2))` before the patch gives the plan similar to the one for `real = decimal` (cross join) because `Analyzer` optimizes expression incorrectly.